lib/modules.nix

```
@@ -412,10 +413,24 @@ rec {
       # modules recursively. It returns the final list of unique-by-key modules
       filterModules = modulesPath: { disabled, modules }:
         let
-          moduleKey = m: if isString m && (builtins.substring 0 1 m != "/")
-            then toString modulesPath + "/" + m
-            else toString m;
-          disabledKeys = map moduleKey disabled;
+          moduleKey = file: m:
+            if isString m && (builtins.substring 0 1 m != "/")
+            then
+              toString modulesPath + "/" + m
+
+            else if isPath m || isString m
+            then toString m
```

I'm just noticing that the previous code allowed any string-coercible value to be used. So we should arguably keep that behavior.

```suggestion
            else if isCoercibleToString m
            then toString m
```

I guess the previous string-coercing behavior wouldn't have worked for your use-case though, because neither `outPath` nor `__toString` are allowed attribute names in modules.

---

I've added a test for this, refactoring the code a bit and rejecting ambiguities between `toString` and `.key`.

Also fyi that function will enter deprecation in summer.

```
  isCoercibleToString = lib.warnIf (lib.isInOldestRelease 2305)
    "lib.strings.isCoercibleToString is deprecated in favor of either isStringLike or isConvertibleWithToString. Only use the latter if it needs to return true for null, numbers, booleans and list of similarly coercibles."
    isConvertibleWithToString;

```